t/semaphore.t

```diff
@@ -0,0 +1,1068 @@
+                sem2:wait(10)
+            end
+            for i=1,#co_array do
+                ngx.thread.wait(co_array[i])
+            end
+            for i=1,g.log_i do
+                --ngx.say(g.log[i].op)
+                if g.log[i].op == 0 then
+                    ngx.say("wait")
+                elseif g.log[i].op == 1 then
+                    ngx.say("post")
+                else
+                    ngx.say("post_all")
+                end
+
+                if g.log[i].state then
+                    ngx.say(g.log[i].state)
+                    if g.log[i].op == 0 then
+                        ngx.say(g.log[i].wait_id)
+                        ngx.say(g.log[i].wake_id)
+                    end
+                else
+                    ngx.say(g.log[i].err)
+                end
+            end
+        ';
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+post
+true
+post
+true
+wait
+true
+1
+1
+wait
+true
+2
+2
+--- no_error_log
+[error]
+
+
+
+=== TEST 21: benchmark
+--- http_config eval: $::HttpConfig
+--- config
+    location = /test {
+        access_log off;
+        content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            if g.id == nil then
+                g.id = 0
+                g.map = {}
+            end
+            local id = g.id
+            g.id = g.id + 1
+            local sem = semaphore.new(0)
+            g.map[id] = sem
+            local res1, res2 = ngx.location.capture_multi{
+                { "/sem_wait",{ method = ngx.HTTP_POST, body = tostring(id) }},
+                { "/sem_post",{ method = ngx.HTTP_POST,body = tostring(id)}},
+            }
+            g.map[id] = nil
+            ngx.say("ok")
+        ';
+    }
+
+    location /sem_wait {
+        content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            ngx.req.read_body()
+            local body = ngx.req.get_body_data()
+            local sem = g.map[tonumber(body)]
+            local ok, err = sem:wait(10)
+            if ok then
+                --ngx.print("wait")
+                ngx.exit(200)
+            else
+                ngx.log(ngx.ERR, err)
+                ngx.exit(500)
+            end
+        ';
+    }
+
+    location /sem_post {
+        content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            ngx.req.read_body()
+            local body = ngx.req.get_body_data()
+            local sem = g.map[tonumber(body)]
+            local ok, err = sem:post()
+            if ok then
+                --ngx.print("post")
+                ngx.exit(200)
+            else
+                ngx.log(ngx.ERR, err)
+                ngx.exit(500)
+            end
+
+    ';
+    }
+
+--- request
+GET /test
+--- response_body
+ok
+--- no_error_log
+[error]
```

Comment:
Again, we should check if the code paths have actually been exercised since tests may have bugs in themselves. We need much stronger checks here instead of just checking the "ok" output and any "[error]" messages.